File: sdk/py/nself_plugin/context/hasura.py
"""
Hasura metadata API wrapper for install/migrate/uninstall hooks.
Mirrors HasuraHelper in the spec. Uses Hasura Metadata API v3 via httpx.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import httpx


@dataclass
class TrackRelationshipOptions:
    from_table: str
    name: str
    type: str  # "object" | "array"
    to_table: str
    from_column: str
    to_column: str


class HasuraHelper:
    """
    Hasura metadata API wrapper.
    Async methods — use in async hooks (install, migrate, uninstall).
    """

    def __init__(self, admin_secret: str, endpoint: str) -> None:
        self._admin_secret = admin_secret
        self._endpoint = endpoint.rstrip("/")

    async def _post(self, path: str, body: dict[str, Any]) -> Any:
        url = f"{self._endpoint}{path}"
        async with httpx.AsyncClient() as client:
            res = await client.post(
                url,
                json=body,
                headers={
                    "Content-Type": "application/json",
                    "x-hasura-admin-secret": self._admin_secret,
                },
                timeout=30,
            )
        if not res.is_success:
            raise RuntimeError(
                f"hasura: POST {path} failed ({res.status_code}): {res.text}"
            )
        return res.json()

    async def add_table(self, table: str, schema: str = "public") -> None:
        """Track a table in Hasura metadata."""
        await self._post("/v1/metadata", {
            "type": "pg_track_table",
            "args": {"source": "default", "schema": schema, "name": table},
        })

    async def remove_table(self, table: str, schema: str = "public") -> None:
        """Untrack a table from Hasura metadata."""
        await self._post("/v1/metadata", {
            "type": "pg_untrack_table",
            "args": {"table": {"schema": schema, "name": table}, "cascade": False},
        })

    async def track_relationship(self, opts: TrackRelationshipOptions) -> None:
        """Add a relationship between two tracked tables."""
        api_type = (
            "pg_create_object_relationship"
            if opts.type == "object"
            else "pg_create_array_relationship"
        )
        await self._post("/v1/metadata", {
            "type": api_type,
            "args": {
                "source": "default",
                "table": opts.from_table,
                "name": opts.name,
                "using": {
                    "foreign_key_constraint_on": {
                        "table": opts.to_table,
                        "columns": [opts.to_column],
                    }
                },
            },
        })

    async def run_metadata(self, type_: str, args: dict[str, Any]) -> Any:
        """Raw Hasura metadata API request (escape hatch)."""
        return await self._post("/v1/metadata", {"type": type_, "args": args})

    async def run_sql(self, sql: str) -> None:
        """Run a SQL migration via Hasura's run_sql API."""
        await self._post("/v2/query", {
            "type": "run_sql",
            "args": {"source": "default", "sql": sql},
        })

File: sdk/ts/src/context/hasura.ts
/**
 * Hasura metadata API wrapper for install/migrate/uninstall hooks.
 * Mirrors the HasuraHelper concept from the spec: addTable, trackRelationship,
 * runMigration. Uses the Hasura Metadata API v3.
 */

export interface TrackRelationshipOptions {
  fromTable: string
  name: string
  type: 'object' | 'array'
  toTable: string
  fromColumn: string
  toColumn: string
}

export interface HasuraHelper {
  /** Track a new table in Hasura metadata. */
  addTable(schema: string, table: string): Promise<void>

  /** Untrack a table from Hasura metadata. */
  removeTable(schema: string, table: string): Promise<void>

  /** Add a relationship between two tracked tables. */
  trackRelationship(opts: TrackRelationshipOptions): Promise<void>

  /** Run raw Hasura metadata API request (escape hatch). */
  runMetadata(type: string, args: Record<string, unknown>): Promise<unknown>

  /** Run a SQL migration via Hasura's run_sql API. */
  runSql(sql: string): Promise<void>
}

export function createHasuraHelper(adminSecret: string, endpoint: string): HasuraHelper {
  async function post(path: string, body: unknown): Promise<unknown> {
    const url = `${endpoint}${path}`
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-hasura-admin-secret': adminSecret,
      },
      body: JSON.stringify(body),
    })
    if (!res.ok) {
      const text = await res.text()
      throw new Error(`hasura: POST ${path} failed (${res.status}): ${text}`)
    }
    return res.json()
  }

  return {
    async addTable(schema: string, table: string): Promise<void> {
      await post('/v1/metadata', {
        type: 'pg_track_table',
        args: {
          source: 'default',
          schema,
          name: table,
        },
      })
    },

    async removeTable(schema: string, table: string): Promise<void> {
      await post('/v1/metadata', {
        type: 'pg_untrack_table',
        args: {
          table: { schema, name: table },
          cascade: false,
        },
      })
    },

    async trackRelationship(opts: TrackRelationshipOptions): Promise<void> {
      const apiType =
        opts.type === 'object' ? 'pg_create_object_relationship' : 'pg_create_array_relationship'
      await post('/v1/metadata', {
        type: apiType,
        args: {
          source: 'default',
          table: opts.fromTable,
          name: opts.name,
          using: {
            foreign_key_constraint_on: {
              table: opts.toTable,
              columns: [opts.toColumn],
            },
          },
        },
      })
    },

    async runMetadata(type: string, args: Record<string, unknown>): Promise<unknown> {
      return post('/v1/metadata', { type, args })
    },

    async runSql(sql: string): Promise<void> {
      await post('/v2/query', {
        type: 'run_sql',
        args: { source: 'default', sql },
      })
    },
  }
}
